Adds new sample "find-obsolete-m365-groups" (#6250)
pnp/cli-microsoft365: docs/docs/sample-scripts/entra/find-obsolete-m365-groups/index.mdx

      $groupInfo.TestStatus = "🟢 OK"

      Write-Host "☀︎ $($groupInfo.Reference.displayName)"

      # Tests
      Test-GroupMembership -Group $groupInfo
      Test-SharePointActivity -Group $groupInfo
      Test-ConversationActivity -Group $groupInfo

      # Report
      New-Report -Group $groupInfo
    }
    
    #Give feedback to user
    Write-Host "`n-------------------------------------------------------------------"
    Write-Host "`SUMMARY" -ForegroundColor DarkGreen
    Write-Host "`-------------------------------------------------------------------"
    Write-Host "`n👉 Found $($Global:ObsoleteGroups.Count) group$($Global:ObsoleteGroups.Count -gt 1 ? 's' : '') with possibly low activity."
    Write-Host "`   Please review the report: " -NoNewline
    Write-Host "$($Script:ReportPath)" -ForegroundColor DarkBlue
  }

  function Test-GroupMembership {
    [CmdletBinding()]
    param (
      [Parameter(Mandatory = $true)] [GroupInfo] $Group
    )

    # Original lists
    $users = m365 entra m365group user list --groupId $Group.Reference.id | ConvertFrom-Json
    $owners = m365 entra m365group user list --groupId $Group.Reference.id --role Owner | ConvertFrom-Json

>     $members = m365 entra m365group user list --groupId $Group.Reference.id --role Member | ConvertFrom-Json

Author: Thanks @milanholemans. I totally agee; from a performance point of view, your suggestions is better.
The only reson why I've chosen the "api-way" (by calling three times the `m365 entra m365group user list` command) is that for the role `members`, the api does not return the same (amount of) values as `$users | Where-Object { $_.roles -contains "Member" }` does. That's why I dug into the implementation (and found the reason) but will open a separate issue that covers this circumstance.

>     $members = m365 entra m365group user list --groupId $Group.Reference.id --role Member | ConvertFrom-Json

Reviewer: That is not intentional then. If that's the case, please log a bug for it.

>     $members = m365 entra m365group user list --groupId $Group.Reference.id --role Member | ConvertFrom-Json

Author: Wrote down some lines and will transfer this to the issues list asap: https://gist.github.com/tmaestrini/4a39ff4008bbb971149d7a8a457786a5

>     $members = m365 entra m365group user list --groupId $Group.Reference.id --role Member | ConvertFrom-Json

Author: Bug report open: https://github.com/pnp/cli-microsoft365/issues/6382

    
    # Consider guests as users that also match the $Script:Guests array
    $guests = $users | Where-Object { $_.id -in $Script:Guests.id }

    # Modify the $members list to only contain users that are not in the $owners list

>     $members = Compare-Object $members $owners -PassThru

Reviewer: This piece of code is still failing for all company groups which have no owners and no members.

Author: You're absolutely right. This would be the case, when every group member has left the company. In this case, any comparison would be obsolete.

Reviewer: As mentioned in the previous review. When you have an all company group, it doesn't have any owners or members.

Author: @milanholemans That's why I've implemented this on line 170:
```
if($null -ne $owners -and $null -ne $members) {
  $members = Compare-Object $members $owners -PassThru
}
```
Doesn't it show up in the code?

Reviewer: The last time I reviewed it wasn't there. I'll recheck in my next review.

    $Group.Membership = [ordered] @{
      Owners  = $owners

>       Members = $members

Author: Well, when members will be retrieved as suggested above, then guests are contained in the members' list and definitely have to be subtracted.

>       Members = $members

Author: Will exclude a `guest` when filtering the `members`

>       Members = $members

Author: @milanholemans That's why I've implemented this on line 170:
```
if($null -ne $owners -and $null -ne $members) {
  $members = Compare-Object $members $owners -PassThru
}
```
Doesn't it show up in the code?

      Guests  = $guests
    }

>     if ($owners.Count -eq 1 -and ($members.Count + $guests.Count) -eq 0) {

Reviewer: Why do we ignore groups with no owners? If a group has one owner, and we delete the user from AD, the group will have no owners.

Author: Checked.

      Write-Host "  → potentially obsolete (abandoned group: only 1 owner left)" -ForegroundColor Yellow
      $reason = "Low user count"
      
      $Group.Membership.Status = "Abandoned ($reason)"
      $Group.TestStatus = "🟡 Warning"
      $Group.Reasons += $reason
      
      try {
        $Global:ObsoleteGroups.Add($Group.Reference.id, $Group)
      }

>       catch { Write-Information "Group was already added to the list of potentially obsolete groups" }

Author: The console will not be flooded (log level INFORMATION) – but nevertheless, I'll restore to emtpy catch.

    }
  }

  function Test-SharePointActivity {
    [CmdletBinding()]
    param (
      [Parameter(Mandatory = $true)] [GroupInfo] $Group
    )
    
    $WarningDate = (Get-Date).AddDays(-90)
    
    $spoSite = $Script:TeamSites | Where-Object { $_.GroupId -eq "/Guid($($Group.Reference.id))/" }

>     $spoWeb = m365 spo web get --url $spoSite.Url | ConvertFrom-Json

Author: I will revert it to `m365 spo site list --filter "Url eq '$($spoSite.Url)'"`as suggested originally (needs date time parsing)

    if ($spoWeb.LastItemUserModifiedDate -lt $WarningDate) {
      Write-Host "  → potentially obsolete (SPO last content modified: $($spoWeb.LastItemUserModifiedDate))" -ForegroundColor Yellow
      $reason = "Low SharePoint activity ($($spoWeb.LastItemUserModifiedDate))"

      $Group.SharePointStatus = @{
        Reason = $reason
      }
      $Group.TestStatus = "🟡 Warning"
      $Group.Reasons += $reason

      try {
        $Global:ObsoleteGroups.Add($Group.Reference.id, $Group)
      }
      catch { Write-Information "Group was already added to the list of potentially obsolete groups" }
    }  
  }

  function Test-ConversationActivity {
    [CmdletBinding()]
    param (
      [Parameter(Mandatory = $true)] [GroupInfo] $Group
    )

    $WarningDate = (Get-Date).AddDays(-365)
      
    $conversations = m365 entra m365group conversation list --groupId $Group.Reference.id | ConvertFrom-Json | Where-Object {

>       [datetime]$_.lastDeliveredDateTime -lt $WarningDate

Author: Good catch! I totally agree: a mailbox will be marked as outdated when there are no conversions that are "younger" than one year (`$WarningDate`).

---
tags:
  - provisioning
  - libraries
  - group mailbox
  - governance
  - teams
  - m365 groups
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> # Finding Obsolete Microsoft 365 Groups with PowerShell

Reviewer: ```suggestion
# Finding obsolete Microsoft 365 groups with PowerShell
```

Author: [Tobias Maestrini](https://github.com/tmaestrini)

This script is based on the [original article](https://petri.com/identifying-obsolete-office-365-groups-powershell) written by [Tony Redmond](https://twitter.com/12Knocksinna).

Like any resource within your Microsoft 365 tenant, M365 Groups can become unused over time.

> This routine uses PowerShell with `m365 cli` commands 

Reviewer: ```suggestion
This routine uses PowerShell with CLI for Microsoft 365
```

> - to gather insights about SharePoint file activity within the related SharePoint site,
> - to do a check against conversation items in the group mailbox,
> - to denote the amount of active people (group owners, members and guests) in the group.

Reviewer: Let's format these as sentenses.

```suggestion
- To gather insights about SharePoint file activity within the related SharePoint site.
- To do a check against conversation items in the group mailbox.
- To denote the amount of active people (group owners, members and guests) in the group.
```

    [PSCustomObject] $SharePointStatus
    [PSCustomObject] $MailboxStatus
    [PSCustomObject] $ChatStatus
    [string] $TestStatus
    [string[]] $Reasons
  }

  function Start-Routine {
    # START ROUTINE
    [CmdletBinding()]
    param (
      [Parameter(Mandatory = $false)] [Switch] $KeepConnectionsAlive,
      [Parameter(Mandatory = $false)] [Switch] $KeepOutputPath
    )

    try {
      Initialize-Params
      if ($KeepOutputPath.IsPresent) { Initialize-ExportPath -KeepOutputPath }
      else { Initialize-ExportPath }
      Get-AllM365Groups
      Start-GroupInsightsTests
      
      Write-Host "`n✔︎ Routine terminated" -ForegroundColor Green
      if (!$KeepConnectionsAlive.IsPresent) {
        m365 logout
      }
    }
    catch {
      Write-Error $_.Exception.Message
    }
  }

  function Initialize-Params {
    Write-Host "🚀 Generating report of obsolete M365 groups within your organization"

    # define globals
    $Global:Path
    $Script:ReportPath = $null
    $Script:Groups = @()
    $Global:ObsoleteGroups = [System.Collections.Generic.Dictionary[string, GroupInfo]]::new()

    Write-Output "Connecting to M365 tenant: please follow the instructions."

>     Write-output "IMPORTANT: You'll need to have global admin permissions!`n"

Reviewer: Global reader will be enough, no?
```suggestion
    Write-output "IMPORTANT: You'll need to have at least global reader permissions!`n"
```

    if ((m365 status --output text) -eq "Logged out") {
      m365 login
    }
  }

  function Initialize-ExportPath {
    [CmdletBinding()]
    param (
      [Parameter(Mandatory = $false)] [Switch] $KeepOutputPath
    )

    if (!$KeepOutputPath.IsPresent -or $null -eq $Global:Path) {
      $Script:Path = Read-Host "Set the path to the folder where you want to export the report data as csv file"
    }
    
    $TestPath = Test-Path -Path  $Script:Path

>     $tStamp = (Get-Date).ToString("yyyyMMddHHmmss")

Reviewer: Let's make it a bit more readable
```suggestion
    $tStamp = (Get-Date).ToString("yyyy-MM-dd HH:mm:ss")
```

    if ($TestPath -ne $true) {
      New-Item -ItemType directory -Path $Script:Path | Out-Null
      Write-Host "Will create file in $($Script:Path): M365GroupsReport-{current date}.csv" -ForegroundColor Yellow
    }
    else {
      Write-Host "Following report file will be created in $($Script:Path): 'M365GroupsReport-$($tStamp).csv'."
      Write-Host "`nAll data will be exported to $($Script:Path): M365GroupsReport-$($tStamp).csv." -ForegroundColor Blue
      Write-Host "Do not edit this file during the scan." -ForegroundColor Blue
    }
    $Script:ReportPath = "$($Script:Path)/M365GroupsReport-$($tStamp).csv"
  }

  function Get-AllM365Groups {
    $groups = m365 entra m365group list --includeSiteUrl | ConvertFrom-Json
    $Script:Groups = $groups | Where-Object { $null -ne $_.siteUrl }
  }

  function Start-GroupInsightsTests {

>     Write-Host "Checking your $($Script:Groups.Count) groups for activity"

Reviewer: ```suggestion
    Write-Host "Checking $($Script:Groups.Count) groups for activity"
```

      $groupInfo.TestStatus = "🟢 OK"

      Write-Host "☀︎ $($groupInfo.Reference.displayName)"

      # Tests
      Test-GroupMembership -Group $groupInfo
      Test-SharePointActivity -Group $groupInfo
      Test-ConversationActivity -Group $groupInfo

      # Report
      New-Report -Group $groupInfo
    }
    
    #Give feedback to user
    Write-Host "`n-------------------------------------------------------------------"
    Write-Host "`SUMMARY" -ForegroundColor DarkGreen
    Write-Host "`-------------------------------------------------------------------"
    Write-Host "`n👉 Found $($Global:ObsoleteGroups.Count) group$($Global:ObsoleteGroups.Count -gt 1 ? 's' : '') with possibly low activity."
    Write-Host "`   Please review the report: " -NoNewline
    Write-Host "$($Script:ReportPath)" -ForegroundColor DarkBlue
  }

  function Test-GroupMembership {
    [CmdletBinding()]
    param (
      [Parameter(Mandatory = $true)] [GroupInfo] $Group
    )

    $users = m365 entra m365group user list --groupId $Group.Reference.id | ConvertFrom-Json
    $owners = $users | Where-Object { $_.userType -eq "Owner" }
    $members = $users | Where-Object { $_.userType -eq "Member" }

>     $guests = $users | Where-Object { $_.userType -eq "Guest" }

Reviewer: The property `userType` is going to be removed in the next major release (https://github.com/pnp/cli-microsoft365/issues/5558). It was never our intention to create this property. People should use `roles` to check if the user is a member or an owner.
Note that `Guest` is not a role. If you want to know if a user is a guest, I think your best shot will be to run `entra user list` and check there if the user is a guest.

Author: Thanks for this notice, although it's a bit confusing: 
in the documented list of examples for `entra m365group user list` (https://pnp.github.io/cli-microsoft365/cmd/entra/m365group/m365group-user-list#examples), you find this: `m365 entra m365group user list --groupDisplayName Developers --filter "userType eq 'Guest'"`. In this case, we should remove the example.

As the `Guest` role is now deprecated, is it correct that to be a `Guest` user, we still can check whether a user is neither `Owner` nor `Member`?

Reviewer: Good point @tmaestrini, it seems like we overlooked removing the example.
You can check if a user is member of owner indeed, but we'll have to check the `roles` property for that. If you want to know if a user is a guest, you'll have to check for that user in Microsoft Entra ID (`entra user get`) for example.
But for me, it's not required to do this check. If you're fine implementing it, go ahead, but it's not required 😊

Author: I've changed the implementation to distinguish between owners, members and guest according to the `role` property.

      $reason = "Low user count"
      
      $Group.Membership.Status = "Abandoned ($reason)"
      $Group.TestStatus = "🟡 Warning"
      $Group.Reasons += $reason
      
      try {
        $Global:ObsoleteGroups.Add($Group.Reference.id, $Group)
      }
      catch {}
    }
  }

  function Test-SharePointActivity {
    [CmdletBinding()]
    param (
      [Parameter(Mandatory = $true)] [GroupInfo] $Group
    )

    function Get-ParsedDate {
      [CmdletBinding()]
      param (
        [Parameter(Mandatory = $true)] [String] $JavascriptDateString
      )

      $dateParts = [regex]::Matches($JavascriptDateString, '\d+') | ForEach-Object { $_.Value }

      # Convert the parts to integers
      $year = [int]$dateParts[0]
      $month = [int]$dateParts[1] + 1
      $day = [int]$dateParts[2]
      $hour = [int]$dateParts[3]
      $minute = [int]$dateParts[4]
      $second = [int]$dateParts[5]
      # $millisecond = [int]$dateParts[6]

      # return a DateTime object
      $dateObject = New-Object -TypeName DateTime -ArgumentList $year, $month, $day, $hour, $minute, $second
      $dateObject
    }
    
    $WarningDate = (Get-Date).AddDays(-90)
    

>     # Not possible to retrieve property 'LastContentModifiedDate' via command 'm365 spo site get --url $group.siteUrl', so we need to use filtering:

Reviewer: Shouldn't this comment be placed a few lines lower?

Author: Sorry for this confiusion, I've had to remove the following line that was commented out.

      Members                   = $Group.Membership.Members.Count
      Guests                    = $Group.Membership.Guests.Count
      "Group Status"            = $Group.Membership.Status ?? "Normal"
      Description               = $Group.Reference.description
      "Conversation Status"     = $Group.MailboxStatus.Reason ?? "Normal"
      "Number of Conversations" = $Group.MailboxStatus.OutdatedConversations ? $Group.MailboxStatus.OutdatedConversations.Length : "n/a"
      "Last Conversation"       = $Group.MailboxStatus.OutdatedConversations ? $Group.MailboxStatus.OutdatedConversations[0].lastDeliveredDateTime : ""
      "Team enabled"            = $Group.Reference.resourceProvisioningOptions -contains 'Team' ? "True" : "False"
      "SPO Status"              = $Group.SharePointStatus.Reason ?? "Normal"
      "SPO Activity"            = $Group.SharePointStatus ? "Low / No document library usage" : "Document library in use"
      "Number of warnings"      = $Group.Reasons.Count
      Status                    = $Group.TestStatus
    }

    $exportObject | Export-Csv -Path $Script:ReportPath -Append -NoTypeInformation
  }

  # START the report generation

>   Start-Routine #-KeepConnectionsAlive -KeepOutputPath

Reviewer: Seems like I'm unable to get the script working. It keeps crashing while checking all groups:

![image](https://github.com/user-attachments/assets/2f1d4d15-6eb8-43d6-9a31-30dee9a156b3)

Author: I will dig into this within the next few days, @milanholemans – I'm already a bit tired... 😃

Author: Can you provide some more details on this error, @milanholemans?

Reviewer: I haven't really dug deeper into it. You're not experiencing this error?

Author: No, no issues so far... 🤷‍♂️

Reviewer: I'll try to look into it ASAP.

Author: Thank you very much!

      Test-ConversationActivity -Group $groupInfo

      # Report
      New-Report -Group $groupInfo
    }
    
    #Give feedback to user
    Write-Host "`n-------------------------------------------------------------------"
    Write-Host "`SUMMARY" -ForegroundColor DarkGreen
    Write-Host "`-------------------------------------------------------------------"
    Write-Host "`n👉 Found $($Global:ObsoleteGroups.Count) group$($Global:ObsoleteGroups.Count -gt 1 ? 's' : '') with possibly low activity."
    Write-Host "`   Please review the report: " -NoNewline
    Write-Host "$($Script:ReportPath)" -ForegroundColor DarkBlue
  }

  function Test-GroupMembership {
    [CmdletBinding()]
    param (
      [Parameter(Mandatory = $true)] [GroupInfo] $Group
    )

    # Original lists
    $users = m365 entra m365group user list --groupId $Group.Reference.id | ConvertFrom-Json
    $owners = m365 entra m365group user list --groupId $Group.Reference.id --role Owner | ConvertFrom-Json
    $members = m365 entra m365group user list --groupId $Group.Reference.id --role Member | ConvertFrom-Json
    
    # Consider guests as users that are not in the $members list

>     $guests = Compare-Object -ReferenceObject $users $members -PassThru

Author: Ok! Thanks for pointing that out, @milanholemans. I didn't know about this behaviour when having 'everyone m365 groups' in the tenant. (At least for me) this is a special behavior that I would not have expected – I suspect that this is due to the fact that 'everyone' points directly to an 'all users' group in Entra ID, since the maximum number of users per group is limited. I will look into this though.

But: shouldn't we also include in the documentation (https://pnp.github.io/cli-microsoft365/cmd/entra/group/group-user-list/) that if group is an Viva engage 'all company group', the return values for `Owner` and `Member` are empty arrays?

>     $guests = Compare-Object -ReferenceObject $users $members -PassThru

Author: Exactly, guests cannot be owner of a group unless they are referenced as user type `Member` in Entra ID. But the `Owner` role has no significance for the validation for the guest users anyway. 

I wanted to realize guest retrieval like this: The command `m365 entra m365group user list` retrieves all members of a group (also including guests). By adding the param `--role Member`, all guest users are filtered out. That's why I wanted to compare those two lists – and the difference between those two lists MUST be the filtered out members: the guest users.

Your suggestion to check the UPN for the string you mentioned (`#EXT#@`) is somewhat unreliable. Since the UPN of every user object in Entra ID is mutable, a system administrator could change this string in any case. Then the guest retrieval query would always return an empty result set. For that reason, I would never do a query on UPN in scripts that must have general validity. 

Do you agree, @milanholemans or am I being too pessimistic? 😃

>     $guests = Compare-Object -ReferenceObject $users $members -PassThru

Reviewer: Tbh I don't think it's worth mentioning it. We also don't want to clutter the docs with too many remarks.

>     $guests = Compare-Object -ReferenceObject $users $members -PassThru

Reviewer: > I wanted to realize guest retrieval like this: The command m365 entra m365group user list retrieves all members of a group (also including guests). By adding the param --role Member, all guest users are filtered out. That's why I wanted to compare those two lists – and the difference between those two lists MUST be the filtered out members: the guest users.

That's not really true. A group consists of owners and members. When you specify `--role member`, you just list all members of the group, including guests since these are just Entra users that can be a guest from a group. The `role` parameter in this command has nothing to do with the user type.

> Your suggestion to check the UPN for the string you mentioned (#EXT#@) is somewhat unreliable. Since the UPN of every user object in Entra ID is mutable, a system administrator could change this string in any case. Then the guest retrieval query would always return an empty result set. For that reason, I would never do a query on UPN in scripts that must have general validity.

You can change it, yes, but that generally speaking not a good idea. If you want to be 100% sure, as mentioned before, we should retrieve all users and check the user type for each user to see if its a guest or not.

>     $guests = Compare-Object -ReferenceObject $users $members -PassThru

Author: Thanks for your patience and comments! 🙌 I will proceed as you told me.

      $reason = "Low user count"
      
      $Group.Membership.Status = "Abandoned ($reason)"
      $Group.TestStatus = "🟡 Warning"
      $Group.Reasons += $reason
      
      try {
        $Global:ObsoleteGroups.Add($Group.Reference.id, $Group)
      }
      catch {}
    }
  }

  function Test-SharePointActivity {
    [CmdletBinding()]
    param (
      [Parameter(Mandatory = $true)] [GroupInfo] $Group
    )

    function Get-ParsedDate {
      [CmdletBinding()]
      param (
        [Parameter(Mandatory = $true)] [String] $JavascriptDateString
      )

      $dateParts = [regex]::Matches($JavascriptDateString, '\d+') | ForEach-Object { $_.Value }

      # Convert the parts to integers
      $year = [int]$dateParts[0]
      $month = [int]$dateParts[1] + 1
      $day = [int]$dateParts[2]
      $hour = [int]$dateParts[3]
      $minute = [int]$dateParts[4]
      $second = [int]$dateParts[5]
      # $millisecond = [int]$dateParts[6]

      # return a DateTime object
      $dateObject = New-Object -TypeName DateTime -ArgumentList $year, $month, $day, $hour, $minute, $second
      $dateObject
    }
    
    $WarningDate = (Get-Date).AddDays(-90)
    

>     # Not possible to retrieve property 'LastContentModifiedDate' via command 'm365 spo site get --url $group.siteUrl', so we need to use filtering:

Author: Yes, we will. 
The comment is related to the call `m365 spo site list --filter` that we make use of for the only reason to have access to the property `LastContentModifiedDate`.

>     # Not possible to retrieve property 'LastContentModifiedDate' via command 'm365 spo site get --url $group.siteUrl', so we need to use filtering:

Reviewer: Yes, but we're not really using `spo site get` like the comment mentions right?

    $spoSite = m365 spo site list --filter "Url -eq '$($Group.Reference.siteUrl)'" | ConvertFrom-Json
    $spoSite.LastContentModifiedDate = Get-ParsedDate -JavascriptDateString $spoSite.LastContentModifiedDate

>     if ($spoSite.LastContentModifiedDate -lt $WarningDate) {

Author: How can I get access to this property? I can't find it my site properties...

>     if ($spoSite.LastContentModifiedDate -lt $WarningDate) {

Reviewer: I was a bit mistaken about the property you were using. The property I'm talking about can be found when using `spo web get` if I'm not mistaken. The downside is that you'll need access to the site which isn't that trivial. So let's maybe just ignore this comment.

>     if ($spoSite.LastContentModifiedDate -lt $WarningDate) {

Author: Got it! Thanks for the hint. I've implemented it, but had to iterate over each site due to the `m365 spo web` command (there's no specific call to retreive all webs at once). But then, we can make use of the `LastItemUserModifiedDate` property.

      Write-Host "  → potentially obsolete (SPO last content modified: $($spoSite.LastContentModifiedDate))" -ForegroundColor Yellow
      $reason = "Low SharePoint activity ($($spoSite.LastContentModifiedDate))"

      $Group.SharePointStatus = @{
        Reason = $reason
      }
      $Group.TestStatus = "🟡 Warning"
      $Group.Reasons += $reason

      try {
        $Global:ObsoleteGroups.Add($Group.Reference.id, $Group)
      }
      catch {}  
    }  
  }

  function Test-ConversationActivity {
    [CmdletBinding()]
    param (
      [Parameter(Mandatory = $true)] [GroupInfo] $Group
    )

    $WarningDate = (Get-Date).AddDays(-365)
      

>     $a = m365 entra m365group conversation list --groupId $Group.Reference.id | ConvertFrom-Json | Where-Object {

Reviewer: Variable $a is not really a clear name.

      [datetime]$_.lastDeliveredDateTime -lt $WarningDate
    }
    if (!$a -or $a.Length -eq 0) { return }

    Write-Host "  → potentially obsolete ($($a.Length) conversation item$($a.Length -gt 1 ? 's' : '') created more than 1 year ago)" -ForegroundColor Yellow
    $reason = "$($a.Length) conversation item$($a.Length -gt 1 ? 's' : '') created more than 1 year ago)"

    $Group.MailboxStatus = @{
      OutdatedConversations = $a | Sort-Object -Property lastDeliveredDateTime
      Reason                = $reason
    }
    $Group.TestStatus = "🟡 Warning"
    $Group.Reasons += $reason

    try {
      $Global:ObsoleteGroups.Add($Group.Reference.id, $Group)
    }

>     catch { }

Reviewer: Noticed you're using `try/catch` to add an item to a list quite some time. Why is this needed?

Author: @milanholemans, this is because I'll mutate the `$Group` object in several steps (by adding testing results). I want to make sure that an possibly obsolete group is only added once to the `ObsoleteGroups` collection. If the group is already existing, the `$Global:ObsoleteGroups.Add(...)` action will throw an error. Instead of always checking if the group was already added to the collection, I'll catch the error and disregard it (because it's intentional behavior).

In these cases, we also could write a debug message to the stdout. Would you prefer this option?

Author: Done!

    catch {
      Write-Error $_.Exception.Message
    }
  }

  function Initialize-Params {
    Write-Host "🚀 Generating report of obsolete M365 groups within your organization"

    # define globals
    $Global:Path
    $Script:ReportPath = $null
    $Script:Groups = @()
    $Script:Guests = @()
    $Script:TeamSites = @()
    $Global:ObsoleteGroups = [System.Collections.Generic.Dictionary[string, GroupInfo]]::new()

    Write-Output "Connecting to M365 tenant: please follow the instructions."
    Write-output "IMPORTANT: You'll need to have at least global reader permissions!`n"
    if ((m365 status --output text) -eq "Logged out") {
      m365 login
    }
  }

  function Initialize-ExportPath {
    [CmdletBinding()]
    param (
      [Parameter(Mandatory = $false)] [Switch] $KeepOutputPath
    )

    if (!$KeepOutputPath.IsPresent -or $null -eq $Global:Path) {
      $Script:Path = Read-Host "Set the path to the folder where you want to export the report data as csv file"
    }
    
    $TestPath = Test-Path -Path  $Script:Path
    $tStamp = (Get-Date).ToString("yyyy-MM-dd HH:mm:ss")
    if ($TestPath -ne $true) {
      New-Item -ItemType directory -Path $Script:Path | Out-Null

>       Write-Host "Will create file in $($Script:Path): M365GroupsReport-$tStamp.csv" -ForegroundColor Yellow

Author: Reverted back…

    
    $Group.MailboxStatus = @{
      NumberOfConversations = $conversations.Length
      LastConversation      = $conversations ? $conversations[0].lastDeliveredDateTime : "n/a"
      OutdatedConversations = 0
      Reason                = ""
    }

    # Return if there are no conversations or the latest conversation is not outdated
    if (!$conversations -or $latestConversation.Count -eq 1) { return }
    
    $outdatedConversations = $conversations | Where-Object {
      [datetime]$_.lastDeliveredDateTime -lt $WarningDate
    }

    Write-Host "  → potentially obsolete ($($outdatedConversations.Length) conversation item$($outdatedConversations.Length -gt 1 ? 's' : '') created more than 1 year ago)" -ForegroundColor Yellow
    $reason = "$($outdatedConversations.Length) conversation item$($outdatedConversations.Length -gt 1 ? 's' : '') created more than 1 year ago"

    $Group.MailboxStatus.OutdatedConversations = $outdatedConversations | Sort-Object -Property lastDeliveredDateTime
    $Group.MailboxStatus.Reason = $reason
    $Group.TestStatus = "🟡 Warning"
    $Group.Reasons += $reason

    try {
      $Global:ObsoleteGroups.Add($Group.Reference.id, $Group)
    }

>     catch { Write-Information "Group was already added to the list of potentially obsolete groups" }

Reviewer: ```suggestion
    catch { }
```

      try {
        $Global:ObsoleteGroups.Add($Group.Reference.id, $Group)
      }

>       catch { Write-Information "Group was already added to the list of potentially obsolete groups" }

Author: change notification output instead of processing empty catch block

    param (
      [Parameter(Mandatory = $true)] [GroupInfo] $Group
    )

> 

Author: delete date parsing due to introduction of new attribute `LastItemUserModifiedDate`instead of `LastContentModifiedDate`

    $WarningDate = (Get-Date).AddDays(-90)
    

>     $spoSite = $Script:TeamSites | Where-Object { $_.GroupId -eq "/Guid($($Group.Reference.id))/" }

Author: change site retrieval to make the script more performant

    $WarningDate = (Get-Date).AddDays(-365)
      

>     $conversations = m365 entra m365group conversation list --groupId $Group.Reference.id | ConvertFrom-Json | Where-Object {

Author: Change variable name to make it clearer
